<?xml version="1.0" encoding="utf-8"?>
<launch>
  <arg name="headless" default="false" />

  <arg name="type" default="quad" />

  <arg name="discrete_path_search_method_name" default="differential_kinematics" />
  <arg name="motion_type"  default="1" /> <!--SE2: 0, SE3: 1 -->

  <arg name="start_squeeze_path_from_real_state" default="true" />

  <arg name="discrete_path_debug_flag" default="false" />
  <arg name="teleop_flag"  default="false" />
  <arg name="path_tf_debug" default= "false" />

  <arg name="save_squeeze_path_flag" default="false" />
  <arg name="load_squeeze_path_flag" default="false" />
  <arg name="path_file_name" default="dragon_planning_log.txt" />

  ###########  Robot Model  ###########
  <group unless="$(arg headless)">
    <include file="$(find dragon)/launch/robot_model.launch" >
      <arg name="headless" value="true" />
      <arg name="need_joint_state" value="false"/>
      <arg name="node_suffix" value="planning"/>
      <arg name="direct_model" value="True" />
      <arg name="direct_model_name" value="$(find squeeze_navigation)/robots/dragon/quad.urdf.xacro"/>
    </include >

    <node name="rviz" pkg="rviz" type="rviz"
    args="-d $(find sampling_based_method)/config/planning.rviz" respawn="false" />
  </group>

  ########## Sampling based method (Moveit) ###########
  <param name="/robot_description_semantic" textfile="$(find sampling_based_method)/config/dragon/$(arg type)/moveit.srdf" />

  ###########  Path Planning and Navigation  ###########
  <node pkg="squeeze_navigation" type="squeeze_navigation_node" name="squeeze_navigation_node" output="screen">

    <param name="debug_verbose" value="false" />
    <param name="headless" value="$(arg headless)" />

    <param name="teleop_flag" value="$(arg teleop_flag)" />
    <!-- robot model -->
    <rosparam file="$(find dragon)/config/$(arg type)/AttitudeControl.yaml" command="load" />

    <!-- discrete path search -->
    <param name="discrete_path_search_method_name" value="squeeze_navigation/$(arg discrete_path_search_method_name)" />

    <rosparam file="$(find squeeze_navigation)/config/$(arg discrete_path_search_method_name)/dragon_$(arg type).yaml" command="load" />
    <rosparam file="$(find squeeze_navigation)/config/$(arg discrete_path_search_method_name)/gap.yaml" command="load"/>

    <param name="discrete_path_debug_flag" value="$(arg discrete_path_debug_flag)" />
    <param name="planning_mode" value="2" /> <!-- if="$(arg transformation)" -->
    <param name="motion_type" value="$(arg motion_type)" />
    <param name="gap_type" value="1" /> <!-- unless="$(arg horizontal_gap)" -->
    <param name="file_state_offset_x" type="double" value="0.0" />
    <param name="file_state_offset_y" type="double" value="0.0" />
    <param name="file_state_offset_z" type="double" value="0.0" />

    <!-- sampling based method -->
    <param name="save_squeeze_path_flag" value="$(arg save_squeeze_path_flag)" />
    <param name="load_squeeze_path_flag" value="$(arg load_squeeze_path_flag)" />
    <param name="start_squeeze_path_from_real_state" value="$(arg start_squeeze_path_from_real_state)" />
    <param name="path_tf_debug" value="$(arg path_tf_debug)" />
    <param name="file_name" value="$(find sampling_based_method)/path_log/dragon/$(arg path_file_name)" />
    <param name="solving_time_limit" type="double" value="36000"/>
    <param name="motion_sequence_rate" value="10"/>

    <!-- RRT -->
    <param name="ompl_mode" value="0"/> <!-- RRT START:0 -->
    <param name="state_validity_check_res" value="0.005"/>
    <param name="valid_segment_count_factor" value="20"/> <!--20-->


    <!-- continuous path generator -->
    <!-- bspline -->
    <param name="bspline_degree" value="5" />

    <!--<rosparam file="$(find squeeze_navigation)/config/end_effector/dragon_$(arg type).yaml" command="load" /> -->
    <rosparam file="$(find squeeze_navigation)/config/end_effector/dragon_$(arg type)_horizontal.yaml" command="load" />
    <!-- Phase3: first approach -->
    <param name="approach_trajectory_period" value="5.0" />

    <!-- Phase5: squeeze -->
    <param name="squeeze_trajectory_period" value="50.0" />

    <!-- navigation -->
    <param name="control_frequency" value="40.0" />
    <param name="joint_control_topic_name" value="/dragon/joints_ctrl" />
    <param name="joint_state_topic_name" value="/dragon/joint_states" />
  </node>

  ######### Debug ##################
  <group if="$(arg path_tf_debug)" >
    <node pkg="dragon" type="dragon_transform_control_node" name="dragon_transformcontrol"  output="screen" respawn="true" >
      <rosparam file="$(find dragon)/config/$(arg type)/AttitudeControl.yaml" command="load" />
      <param name="kinematic_verbose" value="false" />
      <param name="control_verbose" value="false" />
      <param name="debug_verbose" value="false" />
      <param name="real_machine" value="false" />
      <param name="gimbal_control" value="false" />
      <param name="gimbal_control_topic_name" value="/dragon/gimbals_ctrl" />
    </node>
  </group>
</launch>
